Bar width selection via text box

huss opened the pull request:
# Description

The slider for selecting the bar interval is replaced with a text box. The input is validated. The user hits enter to change the graphic to avoid updating the graphic while a number is being entered. The standard buttons and the text box are kept in sync and the corresponding button is highlighted as long as the entered value is one of the standard button choices. Given all of this, the input box is always shown so the button to reveal user input is gone. The max days was changed from 365 to 366 so a full leap year could be in one bar. Finally, minor code changes and comments.

The rationale for this change is:

- You could only see the bar value if holding down the slider where as now it is always visible.
- It was tricky to select a given value with the slider since small movement went several days due to having 365 choices. Using the arrow keys worked but was not intuitive.
- This was the only use of a slider. Input is now consistent with what admin pages do (but regular users will not notice this). It gets rid of a package that had limited use.
- PR #1220 from dependabot indicated the need to upgrade slider as the OED version was several years out of date. See the commit comment on how trying to do this was difficult (at least for me) and this change removes the need for the upgrade as the package is now gone.

## Type of change

- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist

- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request
- [x] You acknowledge that every person contributing to this work has signed the [OED Contributing License Agreement](https://openenergydashboard.github.io/developer/cla.html) and each author is listed in the Description section.

## Limitations

The page styling might be better but it is reasonable.

huss committed:
record of rc-slide

If you update rc-slider 10.6.2 per dependabot PR #1120 it breaks
the code. These changes mostly fix the issues with these exceptions:
1) The tooltip does not work so you  cannot see the value
2) The deprecated functions are not fixed
The rc-slider documentation did not help me fix these.
This commit is designed to be a record before another way
is tried.

huss committed:
remove rc-slider

The replacement is not done yet - only placeholder.

huss committed:
bar interval via text box

The user enters the desired number of days in a text box and then
hits enter. The value is validated by the form. Once input, the bar
graphic is updated with the entered value (if valid). The buttons
with set days are kept in sync with user input.

huss committed:
missed ;

huss committed:
Merge remote-tracking branch 'origin/development' into barWidthSelection

huss committed:
fix fr text

This is not new but the French text for the bar interval help was
missing the start of the text. This is now fixed.

huss committed:
remove core-js package

This package was listed by npm as extraneous so it was pruned.
This removed two babel packages and one other. This appears
to be left over from removing babel or some issue when that
work was merged. It now seems resolved and removes the
warning on install about old version of core-js. It is done here
rather than in an upcomig package update since rc-slider also
needed this.

ChrisMart21 commented:
I agree with the slider going. I feel that typing in a width days is more intuitive. The updates are functional and work as intended. Design wise, I think it would be a nice touch to consider adding another button to the button group so it reads 
day | week | 4 weeks | custom. Then conditionally rendering the input box when selecting custom. What do you think?

huss commented:
> Design wise, I think it would be a nice touch to consider adding another button to the button group so it reads day | week | 4 weeks | custom. Then conditionally rendering the input box when selecting custom. What do you think?

@ChrisMart21 Other than not wanting to do more work this seems like a good suggestion :-) I do have few thoughts that I would like input on:

1. This means four buttons which will make them wider. When the window is narrower or with some languages, this can cause the button(s) on the right to become hidden under the graphic. It already happens a little with three buttons but I decided to leave it. It is a little quicker to be able to hit a button vs a drop down menu but OED has mostly gone to drop down menus for this reason and as we added options. I'm starting to think we should use a drop down here for space issues and consistency. What do you/others think?
2. Currently if you are inputting a bar width in the text box and it matches a standard choice then that bar width button is highlighted. I'm wondering what to do now that custom is selected. Would it stay custom or switch to a standard choice that would hide the custom input? I like the linking but having custom go away is not my favorite. If we choose to start blank each time ( see item 3.) then I'm inclined to leave it as custom. What do you/others think?
3. The text box input changes when a standard bar width is selected. Now that is will be hidden, should it do that or start blank (each time revealed or until a first entry is made)? I'm inclined to start blank every time since the user is likely inputting something different than a standard choice that had to exist and using the arrows in the text box to get there is not a common use case. What do you/others think?

Overall, I'm inclined to use a drop down menu, decouple the buttons from text input where the textbox starts blank each time. As always, thoughts are welcome and encouraged.

ChrisMart21 commented:
I agree with the drop-down for consistency and other mentioned details. I agree, leaving the selection  'custom' while having the input visible. 
However, I feel it'd be better to start with whatever is currently being displayed on the graphic. Essentially keeping the functionality at is now, just hidden unless custom range is desired. I would think knowing where it's at might help the user know more or less where they're at, and how much to shift. That being said I'm unsure which would ' feel' more intuitive until playing with it a bit.

huss committed:
Merge remote-tracking branch 'upstream/development' into barWidthSelection

huss committed:
modify bar duration input

- Dropdown menu rather than buttons.
- Separate choice for custom input whre custom input only shown
if this is selected.
- Make const names follow conventions.
- Added more comments.
- Few other minor changes in style and page layout.
- Removed unused translations.

huss commented:
Thanks to @ChrisMart21  for your suggestions. My recent push should have addressed your comment. I wound up making more changes than expected but the commit lists them. Let me know if you have any thoughts or you can approve with a review if that is appropriate.

ChrisMart21 commented on src/client/app/components/BarControlsComponent.tsx:
```diff
@@ -17,26 +16,81 @@ import TooltipMarkerComponent from './TooltipMarkerComponent';
+			// Set bar duration for graphing, the standard menu value
+			// and hide the custom bar duration input.
+			// Since controlled values know it is a valid integer.
+			console.log('setting barDuration to ', Number(value));
```
Delete Console logs Before Merge.

huss replied:
Thanks. I thought I had removed them all but forgot to do a global search.

ChrisMart21 commented on src/client/app/components/BarControlsComponent.tsx:
```diff
@@ -17,26 +16,81 @@ import TooltipMarkerComponent from './TooltipMarkerComponent';
+			// and hide the custom bar duration input.
+			// Since controlled values know it is a valid integer.
+			console.log('setting barDuration to ', Number(value));
+			updateBarDurationChange(Number(value));
+			setBarDays(value);
+			setShowCustomBarDuration(false);
```
Minor issue when reverting from large bar width days to smaller there can be a bit of lag when plotly renders (from 4 weeks to 1 day for example causes a few seconds of lag on my machine). During this time it is still redering custom input and associated input box. This can be remedied calling updateBardurationChange last, so
```
 setBarDays(...); 
 setShowCustom(...); 
 updateBarDurationChange(...);
```

huss replied:
NIce and fixed. Someday I hope to optimize bar and other DB functions - in this case I suspect using a generated series is slow.

ChrisMart21 reviewed (changes requested):
Overall looks great. The new look blends in quite well with the dropdown instead of buttons. Custom input conditionally rendering is a nice touch.
There a few minor issues as pr comments that should be addressed before a merge. 

Additionally, there is an issue where using a chartlink containing a custom width does not properly display the custom input box on initial render.
Here's a snippet that could help sync with the initial render when this component mounts. There are probably other approaches, but simple.
```tsx
	// should render as custom on initial render if not found in controlled values
	const shouldRenderCustom = !Boolean(['1', '7', '28'].find(days => days == barDuration.asDays().toString()))
	// 
	const [barDays, setBarDays] = React.useState<string>(shouldRenderCustom ? CUSTOM_INPUT : barDuration.asDays().toString());
	const [showCustomBarDuration, setShowCustomBarDuration] = React.useState<boolean>(shouldRenderCustom);
```

huss commented:
> Additionally, there is an issue where using a chartlink containing a custom width does not properly display the custom input box on initial render. 

Good catch. I had thought about if it needed to be dynamic but reasoned that on startup it should be false. I forgot about chart links. I used your code and it works fine now.

huss committed:
address PR comments

huss commented:
Thanks to @ChrisMart21 for the comments. I think I have addressed them all.

huss committed:
add missing semicolon

ChrisMart21 commented on src/client/app/components/BarControlsComponent.tsx:
```diff
@@ -17,26 +16,83 @@ import TooltipMarkerComponent from './TooltipMarkerComponent';
+	const shouldRenderCustom = !(['1', '7', '28'].find(days => days == barDuration.asDays().toString()));
+	// Holds the value of standard bar duration choices used so decoupled from custom and
+	// also to allow special value for custom choice.
+	const [barDays, setBarDays] = React.useState<string>(shouldRenderCustom ? CUSTOM_INPUT : barDuration.asDays().toString());
+	// Holds the value during custom bar duration input so only update graphic when done entering and
+	// separate from standard choices.
+	const [barDaysCustom, setBarDaysCustom] = React.useState<number>(barDuration.asDays());
+	// True if custom bar duration input is active.
+	const [showCustomBarDuration, setShowCustomBarDuration] = React.useState<boolean>(shouldRenderCustom);
 
 	const handleChangeBarStacking = () => {
 		dispatch(graphSlice.actions.changeBarStacking());
 	};
 
-	const handleSliderChange = (value: number) => {
-		setSliderVal(value);
+	// Update when the custom input bar duration is shown/hidden.
+	React.useEffect(() => {
+		if (showCustomBarDuration) {
+			// It changed to true so set the input value as the current barDuration
+			// so shows in custom field at start of input.
+			setBarDaysCustom(barDuration.asDays());
+		}
+		// If false then hiding custom so value does not matter.
+	}, [showCustomBarDuration]);
+
```
Further testing revealed that there was issues keeping redux, and react level state in sync. Similar to the chart link issue, using the history buttons would leave the custom input and dropdown out of sync. My earlier suggestion was useful for syncing state to the initial render, but using the history feature would again leave the component out of sync. 
So moving that logic from initial render into a use effect, will keep in sync when this component's state is altered externally. E.g. not when interacting with the component.
I've tested this, and it works as expected. If you concur this Pr looks good to go. 
```suggestion

	// Holds the value of standard bar duration choices used so decoupled from custom and
	// also to allow special value for custom choice.
	const [barDays, setBarDays] = React.useState<string>(barDuration.asDays().toString());
	// Holds the value during custom bar duration input so only update graphic when done entering and
	// separate from standard choices.
	const [barDaysCustom, setBarDaysCustom] = React.useState<number>(barDuration.asDays());
	// True if custom bar duration input is active.
	const [showCustomBarDuration, setShowCustomBarDuration] = React.useState<boolean>(false);

	const handleChangeBarStacking = () => {
		dispatch(graphSlice.actions.changeBarStacking());
	};

	// Keeps react-level state, and redux state in sync.
	// Two different layers in state may differ especially when externally updated (chart link, history buttons.)
	React.useEffect(() => {
		// Assume Value is valid  since it is comming from state.
		// Do not allow bad values in state.
		const isCustom = !(['1', '7', '28'].find(days => days == barDuration.asDays().toString()));
		setShowCustomBarDuration(isCustom);
		setBarDaysCustom(barDuration.asDays());
		setBarDays(isCustom ? CUSTOM_INPUT : barDuration.asDays().toString());
	}, [barDuration]);
```

ChrisMart21 commented on src/client/app/components/BarControlsComponent.tsx:
```diff
@@ -17,26 +16,83 @@ import TooltipMarkerComponent from './TooltipMarkerComponent';
+	// True if custom bar duration input is active.
+	const [showCustomBarDuration, setShowCustomBarDuration] = React.useState<boolean>(shouldRenderCustom);
 
 	const handleChangeBarStacking = () => {
 		dispatch(graphSlice.actions.changeBarStacking());
 	};
 
-	const handleSliderChange = (value: number) => {
-		setSliderVal(value);
+	// Update when the custom input bar duration is shown/hidden.
+	React.useEffect(() => {
+		if (showCustomBarDuration) {
+			// It changed to true so set the input value as the current barDuration
+			// so shows in custom field at start of input.
+			setBarDaysCustom(barDuration.asDays());
+		}
+		// If false then hiding custom so value does not matter.
+	}, [showCustomBarDuration]);
+
+	// Returns true if this is a valid bar duration.
+	const barDaysValid = (barDays: number) => {
+		return Number.isInteger(barDays) && barDays >= MIN_BAR_DAYS && barDays <= MAX_BAR_DAYS;
 	};
+
+	// Updates values when the standard bar duration menu is used.
+	const handleBarDaysChange = (value: string) => {
+		if (value === CUSTOM_INPUT) {
+			// Set menu value for standard bar to special value to show custom
+			// and show the custom input area.
+			setBarDays(CUSTOM_INPUT);
+			setShowCustomBarDuration(true);
+		} else {
+			// Set the standard menu value, hide the custom bar duration input
+			//  and bar duration for graphing.
+			// Since controlled values know it is a valid integer.
+			setBarDays(value);
+			setShowCustomBarDuration(false);
+			updateBarDurationChange(Number(value));
+		}
+	};
```
Further testing revealed that there was issues keeping redux, and react level state in sync. Similar to the chart link issue, using the history buttons would leave the custom input and dropdown out of sync. My earlier suggestion was useful for syncing state to the initial render, but using the history feature would again leave the component out of sync.
So moving that logic from initial render into a use effect, will keep in sync when this component's state is altered externally. E.g. not when interacting with the component.
I've tested this, and it works as expected. If you concur this Pr looks good to go.

```suggestion
	// Holds the value of standard bar duration choices used so decoupled from custom and
	// also to allow special value for custom choice.
	const [barDays, setBarDays] = React.useState<string>(barDuration.asDays().toString());
	// Holds the value during custom bar duration input so only update graphic when done entering and
	// separate from standard choices.
	const [barDaysCustom, setBarDaysCustom] = React.useState<number>(barDuration.asDays());
	// True if custom bar duration input is active.
	const [showCustomBarDuration, setShowCustomBarDuration] = React.useState<boolean>(false);

	const handleChangeBarStacking = () => {
		dispatch(graphSlice.actions.changeBarStacking());
	};

	// Keeps react-level state, and redux state in sync.
	// Two different layers in state may differ especially when externally updated (chart link, history buttons.)
	React.useEffect(() => {
		// Assume Value is valid  since it is comming from state.
		// Do not allow bad values in state.
		const isCustom = !(['1', '7', '28'].find(days => days == barDuration.asDays().toString()));
		setShowCustomBarDuration(isCustom);
		setBarDaysCustom(barDuration.asDays());
		setBarDays(isCustom ? CUSTOM_INPUT : barDuration.asDays().toString());
	}, [barDuration]);

	// Returns true if this is a valid bar duration.
	const barDaysValid = (barDays: number) => {
		return Number.isInteger(barDays) && barDays >= MIN_BAR_DAYS && barDays <= MAX_BAR_DAYS;
	};

	// Updates values when the standard bar duration menu is used.
	const handleBarDaysChange = (value: string) => {
		if (value === CUSTOM_INPUT) {
			// Set menu value for standard bar to special value to show custom
			// and show the custom input area.
			setBarDays(CUSTOM_INPUT);
			setShowCustomBarDuration(true);
		} else {
			// Set the standard menu value, hide the custom bar duration input
			//  and bar duration for graphing.
			// Since controlled values know it is a valid integer.
			setShowCustomBarDuration(false);
			updateBarDurationChange(Number(value));
		}
	};
```

huss replied:
@ChrisMart21 Thanks. I agree this works. Having found a different history issue a few days ago, I should have thought of this. I guess trying to avoid useEffect code is not a good idea :-). I only edited to remove a line of unused code and minor comment updates. Would you be willing to formally approve the changes so it can be merged without bypassing the rules?

huss committed:
fix state update

This fixes so state is correct even if from history. The suggestion
came from @ChrisMart21.

ChrisMart21 reviewed (approved):
Looks good
